CQT length scale normalization

bmcfee opened the pull request:
This PR implements #412, and a couple of bug-fixes.  Summary of contents:

- Added a `scale` boolean option to all CQT methods.  If enabled, CQT bands are normalized by `sqrt(n[i])` where `n[i]` is the length of the `i`th filter.  This is analogous to `norm='ortho'` mode in `np.fft.fft`.
- Early downsampling is less aggressive by one octave.  Previously, it was too aggressive, and the top-end of each octave was close to nyquist, which resulted in attenuation.
- Magnitude is now continuous and approximately equal for full, hybrid, and pseudo CQT.  This fixes some of the unresolved continuity errors noted in #347.
- Expanded the CQT unit tests to include a white noise continuity test.


Using `scale=True` means that white noise input will look like white noise output in CQT space (flat spectrum).  With `scale=False` (current default), white noise in looks like `1/sqrt(f)`.

`scale=True` makes an impulse look like `sqrt(f)` as opposed to constant for `scale=False`.


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/librosa/librosa/417)
<!-- Reviewable:end -->

bmcfee commented:
The pseudo-cqt with `scale=False` gives a flat response to white noise because it's using a fixed window length for all frequencies.

This causes the previously observed discontinuity in frequency response magnitude of hybrid cqt at the full-pseudo boundary; the compensating factor ought to be something like `n[i] / 2**(np.ceil(np.log2(n[i])))`.

ejhumphrey commented:
*[librosa/core/constantq.py, line 79 at r1](https://reviewable.io:443/reviews/librosa/librosa/417#-KTVh8UZu58k4R1niJAO:-KTVh8UZu58k4R1niJAP:b-aepka7) ([raw file](https://github.com/librosa/librosa/blob/21e7f8693a6d48756a8c4e1f238c935d9215c5f8/librosa/core/constantq.py#L79)):*
> ```Python
>         See `filters.get_window` for details.
> 
>     scale : bool
> ```

how does this affect the choice of `norm=*` in this function? are these independent parameters?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

ejhumphrey commented:
`scale=False` gives a flat response, so this changes the default behavior?

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417#-:-KTVgitMNPoMdrxo5Xnt:b4pvhp6)*
<!-- Sent from Reviewable.io -->

ejhumphrey commented:
*[a discussion](https://reviewable.io:443/reviews/librosa/librosa/417#-KTVhkIXYni4BuLBqg-s:-KTVhkIXYni4BuLBqg-t:b9mi1au) (no related file):*
thoughts on a test that pushes white noise through the CQT with `scale=True/False`? worthwhile or overkill?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

bmcfee commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/librosa/librosa/417#-KTVhkIXYni4BuLBqg-s:-KTVisBFSWlQwHeTuipq:bb6mt9b) (no related file):*
<details><summary><i>Previously, ejhumphrey (Eric J. Humphrey) wrote…</i></summary>
> thoughts on a test that pushes white noise through the CQT with `scale=True/False`? worthwhile or overkill?
</details>
:+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

bmcfee commented:
`scale=False` is the default; behavior should not change.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417#-:-KTVitTrlpiifcObveEr:ba949uk)*
<!-- Sent from Reviewable.io -->

bmcfee commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[librosa/core/constantq.py, line 79 at r1](https://reviewable.io:443/reviews/librosa/librosa/417#-KTVh8UZu58k4R1niJAO:-KTViz6alpeLTmUWVYRL:b-vejd44) ([raw file](https://github.com/librosa/librosa/blob/21e7f8693a6d48756a8c4e1f238c935d9215c5f8/librosa/core/constantq.py#L79)):*
<details><summary><i>Previously, ejhumphrey (Eric J. Humphrey) wrote…</i></summary>
> how does this affect the choice of `norm=*` in this function? are these independent parameters?
</details>
They are independent, yes.

They have to be because the length normalization is different for each octave, but the filter construction is constant across octaves.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

ejhumphrey commented:
*[librosa/core/constantq.py, line 79 at r1](https://reviewable.io:443/reviews/librosa/librosa/417#-KTVh8UZu58k4R1niJAO:-KTVkdK4aIsTpoabf8KK:by6vkdc) ([raw file](https://github.com/librosa/librosa/blob/21e7f8693a6d48756a8c4e1f238c935d9215c5f8/librosa/core/constantq.py#L79)):*
<details><summary><i>Previously, bmcfee (Brian McFee) wrote…</i></summary>
> They are independent, yes.
> 
> They have to be because the length normalization is different for each octave, but the filter construction is constant across octaves.
</details>
so `scale` implies length-scaling only... and the configuration that I would want is `norm=None, scale=True`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

bmcfee commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[librosa/core/constantq.py, line 79 at r1](https://reviewable.io:443/reviews/librosa/librosa/417#-KTVh8UZu58k4R1niJAO:-KTVlx_6_w-G-PEs71KC:b-aj82cd) ([raw file](https://github.com/librosa/librosa/blob/21e7f8693a6d48756a8c4e1f238c935d9215c5f8/librosa/core/constantq.py#L79)):*
<details><summary><i>Previously, ejhumphrey (Eric J. Humphrey) wrote…</i></summary>
> so `scale` implies length-scaling only... and the configuration that I would want is `norm=None, scale=True`?
</details>
No, I think you want `norm=1, scale=True`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

bmcfee commented:
If we disable `scale` in `hybrid_cqt`'s call to `pseudo_cqt`, then things look nicely flat, except for the huge scaling discrepancy:

![image](https://cloud.githubusercontent.com/assets/1190540/19204230/92f2b088-8ca9-11e6-9f2f-77d0e71a2649.png)

bmcfee commented:
Updates: I made a few changes to the re-normalization, but it's now pretty consistently behaving as follows:

![image](https://cloud.githubusercontent.com/assets/1190540/19205472/b68ceef8-8cb0-11e6-9cf5-7031767efb5a.png)

Where the red curve (hybrid, `scale=False`) has a jump discontinuity and scale error by a factor of `sqrt(log(downsample_factor))`.  Still trying to find a clean way to resolve that.

bmcfee commented:
Scale discrepancies are now fixed:

![image](https://cloud.githubusercontent.com/assets/1190540/19207708/158d2976-8cc1-11e6-9819-d333057b0f0f.png)

In the white noise example, the CQT, hybrid cqt, and pseudo-cqt are all basically equal.  The problem was that the early downsampling needs to know whether post-cqt scaling will be applied so that it can correctly compensate.

The test with an impulse is less clean because of the bigger difference between pseudo and full behavior.

I added the `bug` tag to this issue because I think the scale discrepancy we noticed previously in #347 was actually just a bug, not a feature.

bmcfee commented:
While we're at it: should we make `scale=True` the default?  

Opinions from @dpwe @lostanlen @ebattenberg @ejhumphrey ?

(If you're just joining us, this means that white noise in time looks white in CQT, and breaks backwards compatibility.)

bmcfee commented:
2e993f3 reduces the early downsampling nyquist check by one.  I was seeing a lot of scalloping at the top end of the octaves, indicating that we were too close to critical sampling.  This is no longer an issue.

ejhumphrey commented:
regarding the default value of `scale`, I feel defaults should strive to match intuition and take priority over backwards compatibility, and so for me the question is more "what would one expect to get back from a standard CQT?"

that I raised the issue, I would expect a white noise signal to be flat in CQT space (like white noise in an image). It seems like the thing one would want more often, and I (perhaps wrongly) assume filter normalization isn't on the radar of a DSP non-expert.

bmcfee commented:
> "what would one expect to get back from a standard CQT?"

If we treat [Brown'91](https://www.researchgate.net/profile/Christian_Schoerkhuber/publication/228523955_Constant-Q_transform_toolbox_for_music_processing/links/55126aa60cf20bfdad513a3f.pdf) (Eq. 5) as the standard, then it should be length-normalized: divide by `N[k]` for the `k`th basis.

> I would expect a white noise signal to be flat in CQT space

Dividing by `N[k]` actually doesn't get you a white spectrum on white noise if you use `scipy.fftpack.fft` (or API-compatible friends like pyfftw).  You need a `sqrt(N[k])` to achieve this, as demonstrated by the -1/2 slope on the log-log plots above.  AFAIK, other implementations don't do this -- but maybe Ron's / marlib's do?



> I feel defaults should strive to match intuition and take priority over backwards compatibility

I agree with that.  Since we're breaking backwards compatibility anyway to fix the hybrid_cqt bug (and complex output before that), I'm open to changing the default for scaling as well.  But I want to hash out the implications of it before committing to anything.  For example, it will also change how the downstream chroma features behave -- probably for the better?

dpwe commented:
>
>
> I would expect a white noise signal to be flat in CQT space
>
> Dividing by N[k] actually doesn't get you a white spectrum on white noise
> if you use scipy.fftpack.fft (or API-compatible friends like pyfftw). You
> need a sqrt(N[k]) to achieve this, as demonstrated by the -1/2 slope on
> the log-log plots above. AFAIK, other implementations don't do this -- but
> maybe Ron's / marlib's do?
>
To be clear, the FFT implementation isn't relevant here, right?  The only
possible major difference between FFT implementations would be global
scaling (sometimes 1/N in one direction, sometimes 1/sqrt(N) in both
directions), which would show up as step discontinuities each octave if it
was "wrong".  It wouldn't change the CQT output from flat to sloped or vice
versa.


> I feel defaults should strive to match intuition and take priority over
> backwards compatibility
>
> I agree with that. Since we're breaking backwards compatibility anyway to
> fix the hybrid_cqt bug (and complex output before that), I'm open to
> changing the default for scaling as well. But I want to hash out the
> implications of it before committing to anything. For example, it will also
> change how the downstream chroma features behave -- probably for the better?
>

I agree with privileging intuition, but I'm not sure flat CQT is the right
intuition.  Compare these two impulse response images from the earlier
discussion of  CQT <https://github.com/librosa/librosa/issues/341>scaling
(the first one has octave contours, but imagine it smoothly decaying within
each octave block):

[image: Inline image 1][image: Inline image 3]


When you plot it to show the time spread too, having the amplitude decay as
1/N[k]^p looks better, otherwise it seems like the low frequency channels
are indicating way more energy because the high amplitude lasts so much
longer.  Integrating along time in the 1/N[k]^p version seems like each CQ
band delivers the same amount of total energy.

  DAn.


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/librosa/librosa/pull/417#issuecomment-252519361>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAhs0V_Vgp9eCLxHrtcfmevbSJMwJVYdks5qyXIZgaJpZM4KRXo1>
> .
>

bmcfee commented:
> he only
> possible major difference between FFT implementations would be global
> scaling (sometimes 1/N in one direction, sometimes 1/sqrt(N) in both
> directions), which would show up as step discontinuities each octave if it
> was "wrong".

Correct.  The scipy implementation uses a 1/N scaling on the inverse transform.  We already compensate for this when computing each octave's response [here](https://github.com/librosa/librosa/blob/master/librosa/core/constantq.py#L492-L493).



> but I'm not sure flat CQT is the right
> intuition.

This was only intended to refer to white noise input. I think @ejhumphrey's intuition (correct me if I'm wrong) is that white input should have flat spectra at all frames.  As you point out, an impulse should only look flat when time is integrated out, but at should curve upward within a frame.

For an impulse, this is what we currently get:

![image](https://cloud.githubusercontent.com/assets/1190540/19238112/fb2223b6-8ecd-11e6-8807-e1575d0b6fb3.png)

Integrating squared energy over time:

![image](https://cloud.githubusercontent.com/assets/1190540/19238186/576c45de-8ece-11e6-9aef-9e3a85a33066.png)

so the scaled version gets a flat response, like you say, up to the last octave.

The linear slope in the top octave seems a bit suspicious -- I suspect what's happening is that no downsampling is happening here, and the lower octaves are relying on downsampling to implement part of the scaling.  I'll investigate.

dpwe commented:
The impulse responses look lovely, apart from the top octave as you note.

Can you show us the same plots for white noise input?  Or did I miss them
upthread?

Thanks,

  DAn.

On Mon, Oct 10, 2016 at 6:55 AM, Brian McFee <notifications@github.com>
wrote:

> he only
> possible major difference between FFT implementations would be global
> scaling (sometimes 1/N in one direction, sometimes 1/sqrt(N) in both
> directions), which would show up as step discontinuities each octave if it
> was "wrong".
>
> Correct. The scipy implementation uses a 1/N scaling on the inverse
> transform. We already compensate for this when computing each octave's
> response here
> <https://github.com/librosa/librosa/blob/master/librosa/core/constantq.py#L492-L493>
> .
>
> but I'm not sure flat CQT is the right
> intuition.
>
> This was only intended to refer to white noise input. I think @ejhumphrey
> <https://github.com/ejhumphrey>'s intuition (correct me if I'm wrong) is
> that white input should have flat spectra at all frames. As you point out,
> an impulse should only look flat when time is integrated out, but at should
> curve upward within a frame.
>
> For an impulse, this is what we currently get:
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/1190540/19238112/fb2223b6-8ecd-11e6-8807-e1575d0b6fb3.png>
>
> Integrating squared energy over time:
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/1190540/19238186/576c45de-8ece-11e6-9aef-9e3a85a33066.png>
>
> so the scaled version gets a flat response, like you say, up to the last
> octave.
>
> The linear slope in the top octave seems a bit suspicious -- I suspect
> what's happening is that no downsampling is happening here, and the lower
> octaves are relying on downsampling to implement part of the scaling. I'll
> investigate.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/librosa/librosa/pull/417#issuecomment-252627329>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAhs0XuTfdzn-brWI7b7BffxdxLHzjwbks5qykPmgaJpZM4KRXo1>
> .
>

bmcfee commented:
> Can you show us the same plots for white noise input?  Or did I miss them
> upthread?

Those are, more or less, the up-thread plots.  But here's exactly the same code on white input instead of impulse:

![image](https://cloud.githubusercontent.com/assets/1190540/19239178/ddb21250-8ed2-11e6-9c20-119a5ca94b52.png)


![image](https://cloud.githubusercontent.com/assets/1190540/19239182/e2fae4da-8ed2-11e6-9d39-2af623977c36.png)

dpwe commented:
So that's great as it stands IMO.

(My) intuition wants flat responses for both impulses and white noise, but
CQT doesn't work that way (because of the increasing integration window for
lower frequencies).  So we have two settings - one that gives flat
(time-integrated energy for) impulse input, and one that gives flat
response to white noise.

Thanks,

  DAn.

On Mon, Oct 10, 2016 at 7:18 AM, Brian McFee <notifications@github.com>
wrote:

> Can you show us the same plots for white noise input? Or did I miss them
> upthread?
>
> Those are, more or less, the up-thread plots. But here's exactly the same
> code on white input instead of impulse:
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/1190540/19239178/ddb21250-8ed2-11e6-9c20-119a5ca94b52.png>
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/1190540/19239182/e2fae4da-8ed2-11e6-9d39-2af623977c36.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/librosa/librosa/pull/417#issuecomment-252634655>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAhs0TjECIpd4TACsBTkXBqy30JhEyLPks5qyklOgaJpZM4KRXo1>
> .
>

bmcfee commented:
Great.

Any thoughts about what's going on in that top octave?  The ramp on the impulse response seems a little weird to me, but I think it's unavoidable given that it's going to have a flat response in a fixed-window FFT (ie, the one used to compute the top octave), and we're re-scaling it to simulate having variable window lengths.  Since the filters have constant energy but variable length, the peaks will decay with the length, and that's what we're seeing there.

At the lower octaves, this washes out because we're integrating over wider windows and the downsampling smears out the impulse.  (I think?)

Anyway, I think using white noise is a saner test than an impulse for the scaling issue, so I'm inclined to leave it as is.

If folks agree on that, this just leaves the question of what the default behavior should be.

ejhumphrey commented:
so, another signal test that came to me as I was dozing off last night is a frequency modulated sine-sweep, i.e. chirp. The signal would be expressed as `y(t) = α*sin(θ(t) + φ)`. In this instance, with `α` being a constant, I would expect (or perhaps want) the max-amplitude of the chirp in a time-frequency representation to be constant also, i.e. a flat response, regardless of whether the analysis filterbank is linear (DFT) or log-spaced (CQT).

is day-off-Eric being dumb, or...? perhaps it's just that I like to think of the CQT as a pitch-linear complement to the DFT, which would guide my intuition for defaults.

bmcfee commented:
> The signal would be expressed as y(t) = α*sin(θ(t) + φ). In this instance, with α being a constant, I would expect (or perhaps want) the max-amplitude of the chirp in a time-frequency representation to be constant also, i.e. a flat response, regardless of whether the analysis filterbank is linear (DFT) or log-spaced (CQT).

I'm not sure?  It's going to depend on the rate of the sweep.  Lower frequencies will resonate longer than high frequencies with a linear sweep, so it should go down like 1/rate (or something).

bmcfee commented:
Here's MSE of a linear-speed sine sweep from C1 (~ `2**5`) to C8 (~ `2**12`):
```python
sr = 68000
duration = 5
freqs = librosa.note_to_hz(['C1', 'C8']) / sr
y = np.sin(np.cumsum(2 * np.pi * np.linspace(freqs[0],
                                             freqs[-1],
                                             num=duration*sr)))
```
![image](https://cloud.githubusercontent.com/assets/1190540/19247772/c9edb1de-8efa-11e6-8327-4868e79885fe.png)


which, does fit @ejhumphrey's intuition about flat response, modulo setup on the left.  It falls off on the right because the CQT goes one octave higher than the sweep.

dpwe commented:
Any thoughts about what's going on in that top octave? The ramp on the
> impulse response seems a little weird to me, but I think it's unavoidable
> given that it's going to have a flat response in a fixed-window FFT (ie,
> the one used to compute the top octave), and we're re-scaling it to
> simulate having variable window lengths. Since the filters have constant
> energy but variable length, the peaks will decay with the length, and
> that's what we're seeing there.
>
>
I think it's because of the very privileged alignment of the impulse in the
top octave, where it is essentially a single point for all bins.  If you
delayed it by half a sample (i.e., a sinc function, a half-sample delay) it
might look OK.

What is the cutoff frequency of the anti-alias filter before the
downsampling?  Maybe if you LPF the input at 2x that before starting, it
would blur it enough to make it behave the same.

  DAn.


> At the lower octaves, this washes out because we're integrating over wider
> windows and the downsampling smears out the impulse. (I think?)
>
> Anyway, I think using white noise is a saner test than an impulse for the
> scaling issue, so I'm inclined to leave it as is.
>
> If folks agree on that, this just leaves the question of what the default
> behavior should be.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/librosa/librosa/pull/417#issuecomment-252663973>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAhs0cr2ptjmVSBXBNBcOSSdM-SStLprks5qymCHgaJpZM4KRXo1>
> .
>

bmcfee commented:
> I think it's because of the very privileged alignment of the impulse in the
> top octave, where it is essentially a single point for all bins.  If you
> delayed it by half a sample (i.e., a sinc function, a half-sample delay) it
> might look OK.

I gave that a shot: impulse at an odd sample position, then downsampled by 2 with resampy:

![image](https://cloud.githubusercontent.com/assets/1190540/19248906/a8040414-8f00-11e6-8e06-5307c487c6ff.png)

looks pretty much the same.  But I agree without about sample position; if the impulse happens in between frames (frame center + 1/2 hop_length), we get something like:

![image](https://cloud.githubusercontent.com/assets/1190540/19248954/e9618ed6-8f00-11e6-9933-52a1be281c43.png)

This looks right to me: once the signal has been downsampled enough, the spacing of the impulse stops mattering.  Above that threshold, higher frequencies will decay more quickly due to the relative position of the impulse compared to the filter's peaks.

I think I'm convinced that this is all behaving as it should.

bmcfee commented:
So, should I take the responses here as indicating that `scale=True` should be default?  If so, I'll have to make some adjustments to the tests before it's ready for CR.

ejhumphrey commented:
that's been my position, but I am only one vote.

ejhumphrey commented:
*[librosa/core/constantq.py, line 79 at r8](https://reviewable.io:443/reviews/librosa/librosa/417#-KTo25chNYd_bQ1a7fWw:-KTo25chNYd_bQ1a7fWx:b-nubsx) ([raw file](https://github.com/librosa/librosa/blob/7af80cb8cd034640fde66112a62dacbf0240ee78/librosa/core/constantq.py#L79)):*
> ```Python
>         See `filters.get_window` for details.
> 
>     scale : bool
> ```

is it worth mentioning anything about the magnitude response's "flatness" here? or is that too colloquial for respective tastes?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

bmcfee commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[librosa/core/constantq.py, line 79 at r8](https://reviewable.io:443/reviews/librosa/librosa/417#-KTo25chNYd_bQ1a7fWw:-KTo2pI_T6b7Mv8Z_pak:b-kztorb) ([raw file](https://github.com/librosa/librosa/blob/7af80cb8cd034640fde66112a62dacbf0240ee78/librosa/core/constantq.py#L79)):*
<details><summary><i>Previously, ejhumphrey (Eric J. Humphrey) wrote…</i></summary>
> is it worth mentioning anything about the magnitude response's "flatness" here? or is that too colloquial for respective tastes?
</details>
I think that's too specific to be useful in the docstring.  It might make a good example though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

ejhumphrey commented:
*[librosa/core/constantq.py, line 79 at r8](https://reviewable.io:443/reviews/librosa/librosa/417#-KTo25chNYd_bQ1a7fWw:-KTo8ZbfSgLfmTXS8M35:b32v57) ([raw file](https://github.com/librosa/librosa/blob/7af80cb8cd034640fde66112a62dacbf0240ee78/librosa/core/constantq.py#L79)):*
<details><summary><i>Previously, bmcfee (Brian McFee) wrote…</i></summary>
> I think that's too specific to be useful in the docstring.  It might make a good example though.
</details>
+1

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

ejhumphrey commented:
*[librosa/core/constantq.py, line 489 at r8](https://reviewable.io:443/reviews/librosa/librosa/417#-KTo8dNy1RPD-jpkp0H1:-KTo8dNy1RPD-jpkp0H2:b9imuuv) ([raw file](https://github.com/librosa/librosa/blob/7af80cb8cd034640fde66112a62dacbf0240ee78/librosa/core/constantq.py#L489)):*
> ```Python
>     C = fft_basis.dot(D)
> 
>     if scale:
> ```

this is the same block of code as above, right? function worthy?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

ejhumphrey commented:
*[tests/test_constantq.py, line 244 at r8](https://reviewable.io:443/reviews/librosa/librosa/417#-KTo9aRaQZ_Unv2_ORtd:-KTo9aRbdrRgloJ6oZDe:b-b0qjxx) ([raw file](https://github.com/librosa/librosa/blob/7af80cb8cd034640fde66112a62dacbf0240ee78/tests/test_constantq.py#L244)):*
> ```Python
> 
> 
> def test_cqt_white_noise():
> ```

is the sine-sweep test worth folding in here? `|question| >> |suggestion|`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

bmcfee commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[librosa/core/constantq.py, line 489 at r8](https://reviewable.io:443/reviews/librosa/librosa/417#-KTo8dNy1RPD-jpkp0H1:-KToCtPIzkVHjqjhZMur:bmmwihf) ([raw file](https://github.com/librosa/librosa/blob/7af80cb8cd034640fde66112a62dacbf0240ee78/librosa/core/constantq.py#L489)):*
<details><summary><i>Previously, ejhumphrey (Eric J. Humphrey) wrote…</i></summary>
> this is the same block of code as above, right? function worthy?
</details>
It's a small repetition, but I don't think it's worth factoring out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

bmcfee commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[tests/test_constantq.py, line 244 at r8](https://reviewable.io:443/reviews/librosa/librosa/417#-KTo9aRaQZ_Unv2_ORtd:-KToCzdMwxfR5N49FLoX:b1zlyoi) ([raw file](https://github.com/librosa/librosa/blob/7af80cb8cd034640fde66112a62dacbf0240ee78/tests/test_constantq.py#L244)):*
<details><summary><i>Previously, ejhumphrey (Eric J. Humphrey) wrote…</i></summary>
> is the sine-sweep test worth folding in here? `|question| >> |suggestion|`
</details>
I think the white noise test is sufficient.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

ejhumphrey commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417#-:-KToGn4Sj9lS9l8_lJWQ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ejhumphrey commented:
Reviewed 1 of 1 files at r1, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/librosa/librosa/417)*
<!-- Sent from Reviewable.io -->

ejhumphrey commented:
for the record, I still !❤️  reviewable 😞

dpwe commented:
I prefer scale=True (white noise, sine sweep is flat, impulse has lower
amplitude as time support grows).

  DAn.

On Tue, Oct 11, 2016 at 5:48 AM, Brian McFee <notifications@github.com>
wrote:

> So, should I take the responses here as indicating that scale=True should
> be default? If so, I'll have to make some adjustments to the tests before
> it's ready for CR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/librosa/librosa/pull/417#issuecomment-252906898>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAhs0YwGScUvtsqCDouILUxeE0lRSOgeks5qy4WbgaJpZM4KRXo1>
> .
>

bmcfee commented:
Ok!  Thanks @dpwe and @ejhumphrey -- I think this one's ready to go then.
